### PR TITLE
recorder-agent: initial support for multiple recordings

### PR DESCRIPTION
**Pull request by wlozano0collabora**

### CODE REVIEW

File: agent/recorder-agent.c [hunk 1]
```diff
@@ -124,6 +134,15 @@ hwangsae_recorder_agent_send_rest_api (HwangsaeRecorderAgent * self,
   return;
 }
 
+static void
+stream_disconnected_cb (HwangsaeRecorder * recorder,
+    HwangsaeRecorderAgent * self)
+{
+  g_debug ("Stream disconnected, cleaning recorder");
+
+  g_clear_object (&recorder);
```
Comment (wlozano0collabora) [inline agent/recorder-agent.c]: I'm not sure if this is the correct way of cleaning the HwangsaeRecorder. I understand that I need to use a signal as I need to be sure that the pipeline has been cleaned, but maybe I'm missing something.

**Comment by justinjoy** [comment]
In my opinion, it will be better to another separate agent for multiple recordings, rather than adding  features to exist one.

**Comment by wlozano0collabora** [comment]
> In my opinion, it will be better to another separate agent for multiple recordings, rather than adding features to exist one.

If I understand correctly you prefer to have a base `RecorderAgent` class and then have two different implementations, one which supports only one recording and other that supports multiples?

Also, I understand the that the current implementation or `recorder-agent` was only for demo, but the real one should support multiple recordings simultaneously.

Can you confirm/clarify?

**Comment by justinjoy** [comment]
> > In my opinion, it will be better to another separate agent for multiple recordings, rather than adding features to exist one.
> 
> If I understand correctly you prefer to have a base `RecorderAgent` class and then have two different implementations, one which supports only one recording and other that supports multiples?
> 

I don’t think this needs to be a child of existing agent.  But if it is easier than creating new one, it can be a child. The purpose of separation is to provide examples and idea for someone who wants to implement their own ways. 

> Also, I understand the that the current implementation or `recorder-agent` was only for demo, but the real one should support multiple recordings simultaneously.
> 

Yes, that’s right. In real, we need to support recording multiple stream in a process.

**Comment by wlozano0collabora** [comment]
> I don’t think this needs to be a child of existing agent. But if it is easier than creating new one, it can be a child. The purpose of separation is to provide examples and idea for someone who wants to implement their own ways.

Thanks for clarifying. I now get your point. I'll work on this and let you know.

**Comment by wlozano0collabora** [comment]
@justinjoy please check if this approach is OK for you.

I evaluated different approaches as you suggested, but in the end it looked to me more clear to have a base `recorder-agent` which implements the basic stuff without duplicating code and then different flavors of `recorder-agent`.

Of course I can change this if you prefer some other approach.

File: agent/mono-recorder-agent.c [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+#include "config.h"
+
+#include "mono-recorder-agent.h"
+#include <hwangsae/recorder.h>
+#include <glib/gstdio.h>
+#include <gst/gst.h>
+
+#include <glib-unix.h>
+
+struct _HwangsaeMonoRecorderAgent
```
Comment (xhaakon) [inline agent/mono-recorder-agent.c]: May I suggest `HwangsaeSimpleRecorderAgent`? Mono makes a wrong impression it has something to deal with sound.

(I'll continue the review on Monday. Have a nice weekend.)

Comment (wlozano0collabora) [reply agent/mono-recorder-agent.c]: Sure, not problem at all, `HwangsaeSimpleRecorderAgent` sounds nice.

File: agent/recorder-agent.c [hunk 3]
```diff
@@ -51,23 +50,12 @@ struct _HwangsaeRecorderAgent
...
   gchar *recorder_id;
-
-  gboolean is_recording;
-  gint64 recording_id;
-};
+} HwangsaeRecorderAgentPrivate;
 
 /* *INDENT-OFF* */
-G_DEFINE_TYPE (HwangsaeRecorderAgent, hwangsae_recorder_agent, HWANGSAEUL_TYPE_APPLICATION)
+G_DEFINE_TYPE_WITH_CODE (HwangsaeRecorderAgent, hwangsae_recorder_agent, HWANGSAEUL_TYPE_APPLICATION, G_ADD_PRIVATE(HwangsaeRecorderAgent))
```
Comment (xhaakon) [inline agent/recorder-agent.c]: You can `G_DEFINE_TYPE_WITH_PRIVATE` and get rid of the last argument.

File: agent/recorder-agent.c [hunk 4]
```diff
@@ -128,70 +118,16 @@ static gint64
...
-
-  if (!self->is_recording) {
-    g_warning ("recording already stopped");
-    return;
-  }
-
-  self->is_recording = FALSE;
-
-  hwangsae_recorder_stop_recording (self->recorder);
+  g_debug ("Default implementation");
```
Comment (xhaakon) [inline agent/recorder-agent.c:130]: You should probably switch the executable of `hwangsae-recorder.service` to `hwangsae-simple-recorder-agent` until the new default agent gets fully implemented.

Comment (wlozano0collabora) [reply agent/recorder-agent.c:130]: Oh, you are right. I totally missed that part. Thanks for pointing that.

File: debian/hwangsae-recorder.install [hunk 5]
```diff
@@ -1,2 +1,2 @@
-usr/bin/hwangsae-recorder-agent
+usr/bin/hwangsae-simple-recorder-agent
```
Comment (xhaakon) [inline debian/hwangsae-recorder.install:1]: Hmm, I'd install both versions for now.

Comment (wlozano0collabora) [reply debian/hwangsae-recorder.install:1]: But there is just one executable agent version `hwangsae-simple-recorder-agent` which is the simplest and easiest implementation, is supports only one simultaneous recording and it based on `hwangsae_recorder_agent`.

`hwangsae_recorder_agent` is a base recorder agent which implements the common code, such as DBUS API, Lookup, Url, etc. but does not implement `start` nor `stop`

Based on this work I'll later push `hwangsae-mulit-recorder-agent` which will be a new agent which support multiple recording simultaneously, of course base on the common code in `hwangsae_recorder_agent`

Comment (xhaakon) [reply debian/hwangsae-recorder.install:1]: Ah, I see. Yes that makes sense.

**Review by xhaakon — APPROVED** [review]

File: agent/recorder-agent.c (hunk 3 above)
Comment (wlozano0collabora) [reply agent/recorder-agent.c]: Sure.

**Review by justinjoy — APPROVED** [review]
@wlozano0collabora  Could you fix conflict?

**Pull request merged**